### ISSUE TITLE
Refactor CLI into a CLI wrapper and Rust SDK

### ISSUE THREAD

**Issue by fundthmcalculus** [Dentity-Development/dentity-ecosystems-sdk#790]
Per a discussion with @geel9 rust doesn't have an SDK, but the CLI does effectively that. We should refactor these two applications to enable the use of our v2 services from Rust. Separation of concerns here.

- [ ] Add `rust` to the "SDK Change Request" issue template
- [ ] Refactor rust code into a crate
- [ ] Crate publish
- [ ] Make CLI wrapper consume Rust SDK

~Opening a can of worms here, but should we consider moving the CLI into a language other than Rust.~ **Update** we won't do this due to discussion below. @geel9 and I talked about this 2022-06-23, and there are some benefits to doing so:

1. Rust is a slower language to write, less flexible.
2. @tmarkovski is the only strong rust programmer on the team (creates a bus factor problem) / CLI is something we should all be able to support
3. Rust does compile to native binaries, eliminating CLI 3rd party dependencies (.NET, Python, Java)
4. Rust is faster, but do we care for the CLI?
5. .NET is broadly available, and more well known. Caveat is installing .NET core to use it.
6. Golang is native binary, but the lack of support for Windows is major issue
7. Use Rust, but refactor out a Rust SDK separately (see original comments)
8. Other options?

**Comment by janpieterz** [comment]
RE point 5, [we can bundle .NET Core into single executables now](https://docs.microsoft.com/en-us/dotnet/core/deploying/single-file/overview), no need to install anything anymore. If we want that route we should first confirm the Rust SDK can be embedded, as this is exactly something that's potentially a bit more difficult.

**Comment by tmarkovski** [comment]
- [ ] Migrate to clap 3

The likelihood that someone would be integrating Rust into their backend service or front end app is very low. Other than that, I agree with the proposal that the services should be usable as library. At this point, it can be achieved by simply adding a `[lib]` target with some minor changes.

**Comment by fundthmcalculus** [comment]
If .net core allows us to make single binary executables, I would suggest that we switch to dotnet for the c l i.

**Comment by fundthmcalculus** [comment]
J.p., I wasn't suggesting that we would embed the rust SDK inside of the dotnet c l i. Instead, my thought would be to make the CLIA in.net and use the dotnet SDK to do so.

**Comment by janpieterz** [comment]
@fundthmcalculus Gotcha, I thought the CLI embeds the SDK already to do it's work 🤦‍♂️

**Comment by tmarkovski** [comment]
> If .net core allows us to make single binary executables, I would suggest that we switch to dotnet for the c l i.

Curious, what's the benefit of having the CLI in .NET?

**Comment by fundthmcalculus** [comment]
> > If .net core allows us to make single binary executables, I would suggest that we switch to dotnet for the c l i.
> 
> Curious, what's the benefit of having the CLI in .NET?

Rust has a much steeper learning curve, fewer people know it. .NET is more broadly known, so anyone can make changes as needed. If the goal of the CLI is to also have some other functionality, then it would be ideal for the CLI to be in a language/runtime that is more prevalent than Rust. @geel9  can you elaborate further.

**Comment by geel9** [comment]
> > If .net core allows us to make single binary executables, I would suggest that we switch to dotnet for the c l i.
> 
> Curious, what's the benefit of having the CLI in .NET?

I think that .NET is much better suited to this kind of work than Rust is.

From having worked somewhat in the CLI, adding functionality is very cumbersome -- for all the reasons that Rust is a unique and valuable language. 

The CLI doesn't really need baremetal speed, guarantees of memory safety, etc.; we're trading a *lot* of developer comfort in exchange for things which don't really provide much benefit in this context.

Plus, due to the lack of a Rust SDK, we're doing a lot of implementation work for the CLI which could be avoided by just having the CLI use one of our SDKs (like the dotnet one).



Ultimately, I think the CLI should be something that any engineer can feel quite comfortable making changes to (adding a new method, adding an output formatting option, whatever), and working within Rust adds a huge barrier to entry.

Basically: Rust is a systems programming language and the CLI isn't a system.

**Comment by geel9** [comment]
The main advantage of Rust that I see is its ability to export native binaries with minimal hassle. 

If we can get decent native binaries working in .NET (that is, not overly-bloated with runtimes), I'd personally be willing to take on the work to port the CLI.

**Comment by tmarkovski** [comment]
Would you rather do the CLI in Go than in .NET?

The tooling in .NET for executable programs is quite bad. The compilation into a single binary package may be convenient, but there's a lot of downsides, like the size of the binary, the long startup time, etc. Unless a language compiles into native code, it's not a good fit for a CLI. I would actually rather use a python wrapper, and I'm a .NET guy!
I agree, there's a steeper learning curve to Rust than there is for other languages, but we need Rust competency in our team. Our CLI was written by an intern and I think he did an incredible job, I'm sure seasoned engineers will have even easier time.
Most of our SDK is just wrapper around the proto generated services with some custom auth header, I really don't see huge advantage using other languages. If anything, Go might be better suited and seems to be the language of choice for many CLIs.

Little surprised at the pushback here.

**Comment by fundthmcalculus** [comment]
> Would you rather do the CLI in Go than in .NET?

Having worked with go, definitely not. Josh and I discussed golang, and with their objection to treating Windows as a first-class citizen, that's a non starter.

>The tooling in .NET for executable programs is quite bad. The compilation into a single binary package may be convenient, but there's a lot of downsides, like the size of the binary, the long startup time, etc. Unless a language compiles into native code, it's not a good fit for a CLI. I would actually rather use a python wrapper, and I'm a .NET guy!

Say no more. cues up Python CLI. LOL. 😁 That said, you are right about the situation here. I'd do Python given the choice, but I recognize the issues here.

>I agree, there's a steeper learning curve to Rust than there is for other languages, but we need Rust competency in our team. Our CLI was written by an intern and I think he did an incredible job, I'm sure seasoned engineers will have even easier time. Most of our SDK is just wrapper around the proto generated services with some custom auth header, I really don't see huge advantage using other languages. If anything, Go might be better suited and seems to be the language of choice for many CLIs.

Copy. That is a valid requirement, in which case Rust is a totally viable decision for the CLI. I vote then that we keep the CLI in Rust, and refactor it sufficiently to have a Rust SDK wrapper and a Rust CLI. If they are colocated in the repo, that's totally fine.

> Little surprised at the pushback here.

Not pushback, just voicing thoughts and discussion!

**Comment by sethjback** [comment]
> Would you rather do the CLI in Go than in .NET?

I'll always answer yes to that! I'll have to push back a little @fundthmcalculus - there are ton of cross platform CLIs written in golang, especially in the devops world (pretty much the go-to language). The tooling and patterns are fairly mature, including embedded SPAs, which is super cool when you think about the possibility of embedding trinsic studio in the cli. I know we ran into odd SSL issues and windows in our golang SDK. I'm not sure we would have the same issues with a cli.

Having said that - there are multiple mature CLIs in python (e.g. aws and azure) so that is probably a good option too.

Honestly I can't see the argument for using .NET - we would potentially get developer productivity at the expense of usability / performance.

I think a big question is how much more are we wanting to expand the cli and how fast do we want to move? A language other than Rust would probably allow for quicker feature development, but we already have an investment with the current code base so unless we are planning for a large expansion of functionality that we want to move quickly on I'm not sure I see the justification for a rewrite?

**Comment by tmarkovski** [comment]
> ... to have a Rust SDK wrapper and a Rust CLI. If they are colocated in the repo, that's totally fine.

Rust can compile into library or executable by specifying `[lib]` and `[bin]` in it's Cargo.toml file. We currently only specify `[bin]`, so we just need to add library output and it will be usable as SDK. It doesn't need to be separate projects.

**Comment by fundthmcalculus** [comment]
> > Would you rather do the CLI in Go than in .NET?
> 
> I'll always answer yes to that! I'll have to push back a little @fundthmcalculus - there are ton of cross platform CLIs written in golang, especially in the devops world (pretty much the go-to language). The tooling and patterns are fairly mature, including embedded SPAs, which is super cool when you think about the possibility of embedding trinsic studio in the cli.

This would be pretty cool, but not necessarily something useful, per se. That seems to violate the principle of a CLI (command line interface)

> I know we ran into odd SSL issues and windows in our golang SDK. I'm not sure we would have the same issues with a cli.

Given that the CLI would, by necessity, call our services on windows, we would hit exactly the same issue. While this issue had a clear workaround, golang's stubborn refusal to treat Windows as a viable operating system makes it a nonstarter in my mind.

> Having said that - there are multiple mature CLIs in python (e.g. aws and azure) so that is probably a good option too.

I wouldn't be opposed to Python as it were, given it's prevalence in the web automation world. The only caveat here is python installations on various machines. `py2exe` exists by virtue of wrapping `python.exe` with a giant data segment. It works, but it makes the 1K python script a 30MB executable.

> Honestly I can't see the argument for using .NET - we would potentially get developer productivity at the expense of usability / performance.

Usability true. Performance, no. The slowest area of the CLI is the carbon based computer at the terminal. Inside a shell script, yes it runs faster, but if we are serializing and deserializing JSON via pipes, that's not exactly fast either. That said, the overhead of making .NET a single-file application probably isn't worth it.

> I think a big question is how much more are we wanting to expand the cli and how fast do we want to move? A language other than Rust would probably allow for quicker feature development, but we already have an investment with the current code base so unless we are planning for a large expansion of functionality that we want to move quickly on I'm not sure I see the justification for a rewrite?

Yup. This is the crux of the question that Josh and I were discussing. How much is the CLI just another wrapper for the services, and how much is it meant to also be it's own useful tool? If just a wrapper, I say leave it in Rust. If we plan to rapidly add new functionality, there is an argument for a rewrite.



> > ... to have a Rust SDK wrapper and a Rust CLI. If they are colocated in the repo, that's totally fine.
> 
> Rust can compile into library or executable by specifying `[lib]` and `[bin]` in it's Cargo.toml file. We currently only specify `[bin]`, so we just need to add library output and it will be usable as SDK. It doesn't need to be separate projects.

Well, way to bury the hatchet. That's good to know as far as usability outside of the CLI. If it really is that simple *touches wood*, we should just do it. 👍

**Comment by geel9** [comment]
> Would you rather do the CLI in Go than in .NET?

I'd prefer Go over Rust, but .NET over everything else -- **assuming** we're talking solely about developer productivity, which we aren't necessarily.

I think Go would be a decent compromise if we were to switch langs.

> The tooling in .NET for executable programs is quite bad. The compilation into a single binary package may be convenient, but there's a lot of downsides, like the size of the binary, the long startup time, etc. Unless a language compiles into native 
code, it's not a good fit for a CLI. I would actually rather use a python wrapper, and I'm a .NET guy! 

Makes a lot of sense. 

> I agree, there's a steeper learning curve to Rust than there is for other languages, but we need Rust competency in our team. Our CLI was written by an intern and I think he did an incredible job, I'm sure seasoned engineers will have even easier time. Most of our SDK is just wrapper around the proto generated services with some custom auth header, I really don't see huge advantage using other languages. If anything, Go might be better suited and seems to be the language of choice for many CLIs.

It's not even so much the learning curve (though that's relevant) as much as it is the fact that *what Rust makes you do is worthless in the context of a CLI*.

Rust forces you to program in a way that allows the compiler to do *a lot*, and that's very valuable -- for performance-intensive applications.

With a CLI, performance just isn't the root concern, so everything Rust forces you to do is just additional cognitive and SLOC load.


That said, your point about needing Rust competency on the team is well taken. I'm a lot more comfortable stumbling around the CLI while learning Rust because it's a simpler application that isn't mission-critical (I'm not going to bring prod / all our SDKs down by making a mistake in the CLI). **If we have an explicit intention with the CLI of using it to bootstrap Rust competency, I'm all for it**. 

Personally, I was coming at this from the perspective of the CLI as a product; if we are optimizing primarily for the CLI being as good as it can be for our end users, I would say we should do it in something like Go. Outside of that, the decision to go with Rust does make sense.

> Little surprised at the pushback here.

A lot of it is mainly just that every single changed/added endpoint has required that we make changes to every SDK as well as the CLI, which is adding a lot of friction to what we do. However, a lot of this friction is going to be relieved once @fundthmcalculus' auto-generation work begins to bear fruit, which will give us more time to wrangle the CLI.

**Comment by sethjback** [comment]
> This would be pretty cool, but not necessarily something useful, per se. That seems to violate the principle of a CLI (command line interface)

I tend to agree, but think this is one of those areas where user experience trumps model purity - there is something really nice about having a single binary I can script against, but also pass a `-web` flag and get a nice UI. This has definitely been the trend over the last several years in the devops world and I really feel the advantage as an end user of the tooling. Single binary to download, run from anywhere (on the file-system), and I'm not locked in to a single usage type.

It is ambitious and not advocating for it now, I just think it would be super cool to have something like that at some point. Definitely sends the "we have premium tools for our users" vibe, and seems like it could be really handy (a dashboard to manage all your ecosystem templates, the governance framework, do credential validation, etc.)

**Comment by tmarkovski** [comment]
> A lot of it is mainly just that every single changed/added endpoint has required that we make changes to every SDK as well as the CLI, which is adding a lot of friction to what we do. However, a lot of this friction is going to be relieved once @fundthmcalculus' auto-generation work begins to bear fruit, which will give us more time to wrangle the CLI.

That's fair. We did commit to providing SDKs for many languages and ended up making one for all of them. However, we should be able to minimize these frictions, as most updates to the SDK are done quickly by copy/pasting existing code. Most of our SDK methods are very similar, and definitely straightforward in terms of request/response parameters. Introducing new endpoint should require minimal intervention in the SDK to publish that endpoint. Automation can help here, of course, and we're going to slow down on the new endpoints anyway.
CLI is slightly different as it requires maintenance and further abstraction of arguments, but I think with few iterations, this can be minimized as well. It's practice.
We have been consciously making a lot of API changes that required us to reflect them in the SDK, and that adds friction, but the quality of the tooling we provide is pretty good.

**Comment by janpieterz** [comment]
@tmarkovski @fundthmcalculus I'll have to write up a description of this beauty, what capabilities did we unlock doing this as a library? I read the above thread but not sure where we are going to use this SDK.

**Comment by janpieterz** [comment]
Customers can now use the CLI as a library in their rust work. No need to publish as a crate, Github references work. We don't want to really advertise the use as it doesn't follow the SDK standards we have in other languages.
